### PR TITLE
BinaryOperatorSpacesFixer - add (un)align configuration options

### PR DESCRIPTION
**Pull request by SpacePossum**

### CODE REVIEW

File: src/Fixer/Operator/BinaryOperatorSpacesFixer.php [hunk 1]
```diff
@@ -54,6 +95,21 @@ public function fix(\SplFileInfo $file, Tokens $tokens)
...
+            $i = 0;
+
+            /** @var AbstractAlignFixerHelper $helper */
+            foreach ($this->alignFixerHelpers as $helper) {
+                if ($tokens->isChanged()) {
+                    $tokens->clearEmptyTokens();
+                }
+
+                $helper->fix($file, $tokens);
+                ++$i;
```
Comment (SpacePossum) [inline src/Fixer/Operator/BinaryOperatorSpacesFixer.php]: todo - delete useless $i

**Comment by keradus** [comment]
Not sure if introducing possibility to align operators inside `BinaryOperatorSpaces` is good idea.
It makes `BinaryOperatorSpaces` more complex. We could unalign all operators in same way, but we cannot do it easily for aligning them.

Even with single fixer one could don't want to unalign `=`, but also don't want to align `=`, as he is using different logic. For now he dont have option to leave `=` as is and don't touch it.

**Comment by SpacePossum** [comment]
So I misread your suggestion to add configuration, no problem.
If I understand correctly I would suggest to apply https://github.com/FriendsOfPHP/PHP-CS-Fixer/pull/2164 to master after all and than merge the unalign and the align counter part fixers and add configuration to those?

**Comment by keradus** [comment]
> So I misread your suggestion to add configuration, no problem.

no no, the configuration should be there

> If I understand correctly I would suggest to apply #2164 to master after all and than merge the unalign and the align counter part fixers and add configuration to those?

for sure unaligning fixers should be combined and then configurable.
not sure what to do with alignment ones, as both of them have very specific logic.

File: src/Fixer/Operator/BinaryOperatorSpacesFixer.php [hunk 2]
```diff
@@ -24,6 +26,45 @@
...
+
+    /**
+     * @var AbstractAlignFixerHelper[]
+     */
+    private $alignFixerHelpers = array();
+
+    /**
+     * {@inheritdoc}
+     */
+    public function configure(array $configuration = null)
```
Comment (keradus) [inline src/Fixer/Operator/BinaryOperatorSpacesFixer.php]: please document the input

File: src/Fixer/Operator/BinaryOperatorSpacesFixer.php [hunk 3]
```diff
@@ -24,6 +26,45 @@
 final class BinaryOperatorSpacesFixer extends AbstractFixer
 {
     /**
+     * @var array<string, bool>
+     */
+    private $configuration;
+
+    private static $defaultConfiguration = array(
+        'unalign_equals' => true,
+        'unalign_double_arrow' => true,
```
Comment (keradus) [inline src/Fixer/Operator/BinaryOperatorSpacesFixer.php]:
```
'align_double_arrow' => false,
'align_equals' => false,
```

**Comment by SpacePossum** [comment]
>Even with single fixer one could don't want to unalign =, but also don't want to align =, as he is using different logic. For now he dont have option to leave = as is and don't touch it.

fixed, by allowing `null` as configuration value of an item (`align_double_arrow` and/or `align_equals`)

**Comment by keradus** [comment]
aaah, that's the thing.

File: src/Fixer/Operator/BinaryOperatorSpacesFixer.php [hunk 4]
```diff
@@ -111,4 +180,21 @@ private function isDeclareStatement(Tokens $tokens, $index)
 
         return false;
     }
+
+    private function runHelperFixers(\SplFileInfo $file, Tokens $tokens)
+    {
+        $helpersToRunCount = count($this->alignFixerHelpers);
+        if (0 === $helpersToRunCount) {
+            return;
+        }
```
Comment (keradus) [inline src/Fixer/Operator/BinaryOperatorSpacesFixer.php]: all 4 lines could be removed, that early return doesn't provide any good here, the same will happen if array is empty

**Comment by keradus** [comment]
Upgrade guide needs to be updated

**Comment by SpacePossum** [comment]
Done...
damn this PR got big, lucky this one is the biggest of merging conflicting Fixers ;)

**Comment by keradus** [comment]
Thank you @SpacePossum.

**Pull request closed without merging**